### PR TITLE
Fix a race condition with git_pillar mountpoints

### PR DESCRIPTION
**Pull request by terminalmage**
When multiple independent jobs are attempting to compile git_pillar at the same time, and they are using mountpoints, this results in the repo linkdir being removed during the compile process, causing some jobs to fail. This PR implements locking to avoid this.

It also updates the cache runner to be able to clear mountpoint locks, and also fixes a regression which I recently introduced into the mainline branch only (not to any release branch).

### CODE REVIEW

**Review by cachedout — COMMENTED** [review]
While I understand that this is technically a bugfix, I'd personally feel better about this in develop.

**Comment by terminalmage** [comment]
@cachedout rebased and head updated to `develop`.

File: salt/runners/cache.py [hunk 1]
```diff
@@ -310,20 +310,28 @@ def clear_git_lock(role, remote=None, **kwargs):
...
-    type : update,checkout
-        The types of lock to clear. Can be ``update``, ``checkout``, or both of
-    et (either comma-separated or as a Python list).
+    type : update,checkout,mountpoint
+        The types of lock to clear. Can be one or more of ``update``,
+        ``checkout``, and ``mountpoint``, and can be passed either as a
+        comma-separated or Python list.
 
         .. versionadded:: 2015.8.8
+        .. versionchanged:: 2017.7.3
```
Comment (rallytime) [inline salt/runners/cache.py]: This should be changed to `Oxygen` since this PR was moved to `develop`.

**Review by rallytime — CHANGES_REQUESTED** [review]
This is nice @terminalmage.

I had one tiny thing that should change, and it looks like there are merge conflicts on this as well.

File: salt/runners/cache.py (hunk 1 above)
Comment (terminalmage) [reply salt/runners/cache.py]: Fixed.

**Review by rallytime — APPROVED** [review]

**Review by cachedout — APPROVED** [review]

**Pull request merged**